### ISSUE TITLE
test_dist_optim is flaky

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#34141]
## 🐛 Bug

`test_dist_optim` in `dist_optimizer_test.py` is flaky: 

```
Mar 03 07:14:18 Running tests...
Mar 03 07:14:18 ----------------------------------------------------------------------
Mar 03 07:14:19 Process process 3:
Mar 03 07:14:19 Traceback (most recent call last):
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
Mar 03 07:14:19     self.run()
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/multiprocessing/process.py", line 99, in run
Mar 03 07:14:19     self._target(*self._args, **self._kwargs)
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 243, in _run
Mar 03 07:14:19     getattr(self, test_name)()
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 68, in new_test_method
Mar 03 07:14:19     return_value = old_test_method(self, *arg, **kwargs)
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/dist_optimizer_test.py", line 200, in test_dist_optim
Mar 03 07:14:19     self.assertEqual(new_w1, module1.get_w())
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 872, in assertEqual
Mar 03 07:14:19     assertTensorsEqual(x, y)
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_utils.py", line 842, in assertTensorsEqual
Mar 03 07:14:19     self.assertLessEqual(max_err, prec, message)
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/unittest/case.py", line 1232, in assertLessEqual
Mar 03 07:14:19     self.fail(self._formatMessage(msg, standardMsg))
Mar 03 07:14:19   File "/opt/python/nightly/lib/python3.7/unittest/case.py", line 680, in fail
Mar 03 07:14:19     raise self.failureException(msg)
Mar 03 07:14:19 AssertionError: tensor(0.1678, grad_fn=<MaxBackward1>) not less than or equal to 1e-05 : 
Mar 03 07:19:20 E..
Mar 03 07:19:20 ======================================================================
Mar 03 07:19:20 ERROR [300.229s]: test_dist_optim (__main__.DistOptimizerTestWithSpawn)
Mar 03 07:19:20 ----------------------------------------------------------------------
Mar 03 07:19:20 Traceback (most recent call last):
Mar 03 07:19:20   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 175, in wrapper
Mar 03 07:19:20     self._join_processes(fn)
Mar 03 07:19:20   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 256, in _join_processes
Mar 03 07:19:20     self._check_return_codes(elapsed_time)
Mar 03 07:19:20   File "/opt/python/nightly/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 275, in _check_return_codes
Mar 03 07:19:20     raise RuntimeError('Process {} terminated or timed out after {} seconds'.format(i, elapsed_time))
Mar 03 07:19:20 RuntimeError: Process 0 terminated or timed out after 300.1386160850525 seconds
Mar 03 07:19:20 
Mar 03 07:19:20 ----------------------------------------------------------------------
Mar 03 07:19:20 Ran 3 tests in 302.366s
Mar 03 07:19:20 
Mar 03 07:19:20 FAILED (errors=1)
Mar 03 07:19:20 
Mar 03 07:19:20 Generating XML reports...
Mar 03 07:19:20 Traceback (most recent call last):
Mar 03 07:19:20   File "test/run_test.py", line 493, in <module>
Mar 03 07:19:20     main()
Mar 03 07:19:20   File "test/run_test.py", line 486, in main
Mar 03 07:19:20     raise RuntimeError(message)
Mar 03 07:19:20 RuntimeError: distributed/rpc/test_dist_optimizer_spawn failed!
Mar 03 07:19:20 + cleanup
Mar 03 07:19:20 + retcode=1
Mar 03 07:19:20 + set +x
```

See https://app.circleci.com/jobs/github/pytorch/pytorch/4668683

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @rohan-varma @xush6528 @jjlilley @osalpekar

**Comment by rohan-varma** [comment]
@aazzolini Any idea on why this may fail? It seems to be very rare so not sure if there's a race condition with the seeding or something like that.

**Comment by pritamdamania87** [comment]
Tried to repro this with 500 runs in OSS and internal and couldn't reproduce this.

**Comment by pritamdamania87** [comment]
Closing this out since we haven't been able to reproduce the failure. We can revisit this if we this consistently failing in the future.

**Comment by ezyang** [comment]
If you cannot reproduce locally, I would suggest thinking about what kind of logging or rage-style reporting could be installed to help you debug from the (rare) CI crashes. This is a doubly good exercise to do for a distributed system since any extra robustness you add here will also help you diagnose production crashes. I'll emphasize that setting up this kind of logging infrastructure is not an easy task; you'll have to think carefully about how to do it performantly... but I find that this always pays big dividends in the end.

**Comment by mrshenli** [comment]
This strikes again

https://app.circleci.com/pipelines/github/pytorch/pytorch/146275/workflows/166d2803-f5e7-4903-8b1b-636ddb505921/jobs/4946796

**Comment by mrshenli** [comment]
@ezyang is it true that if I have multiple threads in a process, they will share the same CPU RNG?

**Comment by ezyang** [comment]
If they're using the global rng, yes, I think so.

**Comment by mrshenli** [comment]
closed by #35425
